### PR TITLE
Add support for JAX to model_1/2a/3a

### PR DESCRIPTION
**Pull request by AaronDJohnson**
Add `jax` support for some models that exist in `enterprise_extensions`. This allows for gradients and automatic vectorization for `model_2a` and `model_3a`. Note that these functions are extremely slow without `jit` compiling them. This makes it difficult to include in `enterprise` properly. `jit` does not currently work on BayesEphem, white noise signals, and advanced noise model signals. This is due to subclassing of numpy arrays (ShermanMorrison, etc.) and standard `jax` control flow issues.

### CODE REVIEW

**Comment by AaronDJohnson** [comment]
I think I've decided on a name for the PR....

**Comment by AaronDJohnson** [comment]
Converted to draft until code freeze is over. Currently, actions are failing due to poor coverage of the new files.

**Review by paulthebaker — COMMENTED** [review]
I have not fully reviewed everything.  However, I can see that this PR adds a ton of duplicate code.

It looks like `jax` replaces some (all?) `numpy` functions.  Could we keep the same code and add a switch to turn on `jax` if requested?  This would hopefully only modify the import area of the code, adding some flag catching and then just `import jax.numpy as np`?  How often do we do things that would break with `jax`, e.g. modifying arrays in place?

**Comment by AaronDJohnson** [comment]
> I have not fully reviewed everything. However, I can see that this PR adds a ton of duplicate code.
> 
> It looks like `jax` replaces some (all?) `numpy` functions. Could we keep the same code and add a switch to turn on `jax` if requested? This would hopefully only modify the import area of the code, adding some flag catching and then just `import jax.numpy as np`? How often do we do things that would break with `jax`, e.g. modifying arrays in place?

There are issues with using flags. The functions have the same names, but sometimes have different parameters (jnp.where, for example). Additionally, `jax` arrays cannot be modified using indices (https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#in-place-updates).

I think it will take some thinking to reduce the amount of repeated code here. I'll continue picking away at it.

**Comment by vallis** [comment]
Amazing work! I do think it belongs in enterprise though, where we could go with an "import jax.numpy as np" strategy and then make strategic modifications to enterprise classes to allow the jax backend. Maybe in a "v4" branch, so we can also prune unused corners of enterprise?

**Comment by AaronDJohnson** [comment]
> Amazing work! I do think it belongs in enterprise though, where we could go with an "import jax.numpy as np" strategy and then make strategic modifications to enterprise classes to allow the jax backend. Maybe in a "v4" branch, so we can also prune unused corners of enterprise?

Thanks, @vallis. I think you're right that this belongs in `enterprise`. I'll start thinking about how best to do this.

**Comment by Hazboun6** [comment]
I also agree with @vallis that this should be part of enterprise. Should we make a `jax` branch of enterprise where we can start working on the necessary changes?
